File: cpp/tests/test_pacman.cpp
#include <gtest/gtest.h>
#include "../pacman.h"
#include "../maze.h"

TEST(PacmanTest, InitialPosition) {
    Pacman pac;
    EXPECT_EQ(pac.getGridX(), 5);
    EXPECT_EQ(pac.getGridY(), 5);
}

TEST(PacmanTest, MoveRight) {
    Pacman pac;
    Maze maze;
    pac.setDirection(1, 0);
    pac.update(maze);
    EXPECT_EQ(pac.getGridX(), 6);
    EXPECT_EQ(pac.getGridY(), 5);
}

TEST(PacmanTest, MoveDown) {
    Pacman pac;
    Maze maze;
    pac.setDirection(0, 1);
    pac.update(maze);
    EXPECT_EQ(pac.getGridX(), 5);
    EXPECT_EQ(pac.getGridY(), 6);
}

TEST(PacmanTest, CannotMoveIntoWall) {
    Pacman pac;
    Maze maze;
    // Move Pacman to position adjacent to a border wall
    // Position (1,1) is adjacent to wall at (0,1)
    // First navigate pacman to (1,1) by moving left several times
    pac.setDirection(-1, 0); // left
    for (int i = 0; i < 4; ++i) {
        pac.update(maze);
    }
    // Now at (1,5) — try to move further left into wall at (0,5)
    pac.setDirection(-1, 0);
    pac.update(maze);
    EXPECT_EQ(pac.getGridX(), 1); // Should not move into wall
}

TEST(PacmanTest, DirectionBuffering) {
    Pacman pac;
    Maze maze;
    pac.setDirection(1, 0); // right
    pac.update(maze);
    EXPECT_EQ(pac.getGridX(), 6);
    // Change to down, should apply on next update
    pac.setDirection(0, 1);
    pac.update(maze);
    EXPECT_EQ(pac.getGridX(), 6);
    EXPECT_EQ(pac.getGridY(), 6);
}

File: python/tests/test_pacman.py
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from maze import Maze
from pacman import Pacman


class TestInitialPosition:
    """Pacman should start at the coordinates passed to the constructor."""

    def test_initial_position(self):
        pacman = Pacman(5, 5, 20)
        assert pacman.grid_x == 5
        assert pacman.grid_y == 5


class TestCannotMoveIntoWall:
    """Setting direction toward a wall and updating should leave Pacman in place."""

    def test_cannot_move_into_wall(self):
        maze = Maze()
        # Start Pacman at (1, 1) -- surrounded by border wall on the left (x=0)
        pacman = Pacman(1, 1, 20)
        pacman.set_direction(-1, 0)  # Try to move left into wall at (0,1)
        pacman.update(maze)
        assert pacman.grid_x == 1, "Pacman should not move into a wall"
        assert pacman.grid_y == 1


class TestCanMoveIntoOpenSpace:
    """Setting direction to an open tile and updating should move Pacman."""

    def test_can_move_into_open_space(self):
        maze = Maze()
        # (5, 5) is open; (6, 5) is also open
        pacman = Pacman(5, 5, 20)
        pacman.set_direction(1, 0)  # Move right
        pacman.update(maze)
        assert pacman.grid_x == 6
        assert pacman.grid_y == 5


class TestDirectionQueuing:
    """Buffered next-direction should be applied when it becomes valid."""

    def test_direction_queuing(self):
        maze = Maze()
        # Start at (1, 1). Moving down is open (1,2), moving left is wall (0,1).
        pacman = Pacman(1, 1, 20)

        # Set current direction to down by updating once
        pacman.set_direction(0, 1)
        pacman.update(maze)
        assert pacman.grid_x == 1
        assert pacman.grid_y == 2  # moved down

        # Now queue right direction (1, 0). (2, 2) is open.
        pacman.set_direction(1, 0)
        pacman.update(maze)
        assert pacman.grid_x == 2, "Queued right direction should be applied"
        assert pacman.grid_y == 2
